ddtrace/contrib/internal/vertexai/_utils.py

```diff
@@ -0,0 +1,302 @@
+            "vertexai.request.contents.%d.text" % content_idx,
+            integration.trunc("[Non-text content object: {}]".format(repr(content))),
+        )
+        return
+    for part_idx, part in enumerate(parts):
+        _tag_request_content_part(span, integration, part, part_idx, content_idx)
+
+
+def _tag_response_part(span, integration, part, part_idx, candidate_idx):
+    """Tag the generation span with response part text and function calls."""
+    text = part.get("text", "")
+    span.set_tag_str(
+        "vertexai.response.candidates.%d.content.parts.%d.text" % (candidate_idx, part_idx),
+        integration.trunc(str(text)),
+    )
+    function_call = part.get("function_call", None)
+    if not function_call:
+        return
+    span.set_tag_str(
+        "vertexai.response.candidates.%d.content.parts.%d.function_call.name" % (candidate_idx, part_idx),
+        integration.trunc(str(function_call.get("name", ""))),
+    )
+    span.set_tag_str(
+        "vertexai.response.candidates.%d.content.parts.%d.function_call.args" % (candidate_idx, part_idx),
+        integration.trunc(str(function_call.get("args", {}))),
+    )
+
+
+def tag_request(span, integration, instance, args, kwargs):
+    """Tag the generation span with request details.
+    Includes capturing generation configuration, system prompt, and messages.
+    """
+    # instance is either a chat session or a model itself
+    model_instance = instance if isinstance(instance, GenerativeModel) else instance._model
+    contents = get_argument_value(args, kwargs, 0, "contents")
+    history = getattr(instance, "_history", [])
+    if history:
+        if isinstance(contents, list):
+            contents = history + contents
+        if isinstance(contents, Part) or isinstance(contents, str) or isinstance(contents, dict):
+            contents = history + [contents]
+    generation_config_dict = get_generation_config_from_model(model_instance, kwargs)
+    system_instructions = get_system_instruction_parts_from_model(model_instance)
+    stream = kwargs.get("stream", None)
+
+    if generation_config_dict is not None:
+        for k, v in generation_config_dict.items():
+            span.set_tag_str("vertexai.request.generation_config.%s" % k, str(v))
+
+    if stream:
+        span.set_tag("vertexai.request.stream", True)
+
+    if not integration.is_pc_sampled_span(span):
+        return
+
+    if system_instructions:
+        for idx, part in enumerate(system_instructions):
+            span.set_tag_str(
+                "vertexai.request.system_instruction.%d.text" % idx, integration.trunc(str(part.text if hasattr(part, "text") else part))
+            )
+
+    if isinstance(contents, str):
+        span.set_tag_str("vertexai.request.contents.0.text", integration.trunc(contents))
+        return
+    elif isinstance(contents, dict):
+        span.set_tag_str("vertexai.request.contents.0.text", integration.trunc(str(contents)))
+        return
```

Comment:
Probably can combine these two conditions and just cast `contents` to str regardless